### ISSUE TITLE
GM storage in unload event handler

### ISSUE THREAD

**Issue by kirill9000** [greasemonkey/greasemonkey#2435]
It works as expected when reloading/navigating the tab, but doesn't work at all on tab closing (GM_getValue always returns 0, GM_deleteValue has no effect). See the script below for demonstration.

Tested on latest (3.9) Greasemonkey with Firefox 49 and 45.4 ESR.

```
// ==UserScript==
// @name    gm_storage_unload_bug_test
// @grant   GM_getValue
// @grant   GM_setValue
// @grant   GM_deleteValue
// @include https://github.com/*
// ==/UserScript==

if (GM_getValue("test")) console.log("FAIL: test should have been removed in onunload by now.");

let now = Date.now();
GM_setValue("test", now);

window.addEventListener("unload", function onunload() {
    window.removeEventListener("unload", onunload, false);
    let value = GM_getValue("test");
    if (value !== now) console.log("FAIL: Expected: " + now + ", got: " + value);
    GM_deleteValue("test")
}, false);
```

**Comment by janekptacijarabaci** [comment]
And what if you replace the event: "unload" => "beforeunload"?
It triggers __before__ the window is closed or changes location.

Otherwise, this message is returned "0":
https://github.com/greasemonkey/greasemonkey/blob/3.9/modules/storageFront.js#L59
This isn't directed nowhere.

**Comment by kirill9000** [comment]
beforeunload works fine. But I wonder if unload could be fixed?

**Comment by janekptacijarabaci** [comment]
> But I wonder if unload could be fixed?

I don't think so.

**Comment by kirill9000** [comment]
I assume this was closed since the new API in v4.0 is asynchronous. But the issue wasn't resolved. Can we use the new API from unload/beforeunload? Is there any guarantee that values would be properly saved?

**Comment by Sxderp** [comment]
> Is there any guarantee that values would be properly saved?
Unfortunately I believe the answer to that is no.
